store/postgres/src/chain_store.rs

```diff
@@ -1813,6 +1841,164 @@ impl ChainStoreTrait for ChainStore {
     }
 }
 
+mod recent_blocks_cache {
+    use super::*;
+    use std::collections::BTreeMap;
+
+    struct CachedBlock {
+        ptr: BlockPtr,
+        data: Option<json::Value>,
+        parent_hash: BlockHash,
+    }
+
+    struct Inner {
+        // Note: At all times, only a continuous range of blocks will be in the
+        // cache, i.e. without gaps. This constraint makes it possible to easily
+        // verify a direct line of ancestry between two blocks via parent hashes.
+        blocks: BTreeMap<BlockNumber, CachedBlock>,
+        // We only store these many blocks.
+        capacity: usize,
+    }
+
+    impl Inner {
+        fn get_block(
+            &self,
+            child: &BlockPtr,
+            offset: BlockNumber,
+        ) -> Option<(&BlockPtr, Option<&json::Value>)> {
+            // Before we can go find the ancestor, we need to make sure that
+            // we're looking for the ancestor of the right block, i.e. check if
+            // the hash (and number) of the child matches.
+            let child_is_cached = &self.blocks.get(&child.number)?.ptr == child;
+
+            if child_is_cached {
+                let ancestor_block_number = child.number - offset;
+                let block = self.blocks.get(&ancestor_block_number)?;
+                Some((&block.ptr, block.data.as_ref()))
+            } else {
+                None
+            }
+        }
+
+        fn chain_head(&self) -> Option<&BlockPtr> {
+            self.blocks.iter().next_back().map(|b| &b.1.ptr)
+        }
+
+        fn last_block(&self) -> Option<&CachedBlock> {
+            self.blocks.iter().next().map(|b| b.1)
+        }
+
+        fn evict_if_necessary(&mut self) {
+            while self.blocks.len() > self.capacity && !self.blocks.is_empty() {
+                let block_num = *self.blocks.iter().next().unwrap().0;
+                self.blocks.remove(&block_num);
+            }
+        }
+
+        fn insert_block(
+            &mut self,
+            ptr: BlockPtr,
+            data: Option<json::Value>,
+            parent_hash: BlockHash,
+        ) {
+            let block = CachedBlock {
+                ptr,
+                data,
+                parent_hash,
+            };
+
+            let chain_head = if let Some(ch) = self.chain_head() {
+                ch
+            } else {
+                // We don't have anything in the cache, so we're free to store
+                // everything we want.
+                self.blocks.insert(block.ptr.number, block);
+                return;
+            };
+
+            if chain_head.number == block.ptr.number - 1 && chain_head.hash == block.parent_hash {
```

Comment:
It might be good to have a utility function to check if one block is the parent of another to make this more readable